pkg/status/combinedstatus-resolver.go

```diff
@@ -0,0 +1,470 @@
+
+	currentSpec := c.statusCollectorNameToSpec[statusCollector.Name]
+	if currentSpec != nil && statusCollectorSpecsMatch(currentSpec, &statusCollector.Spec) {
+		return nil // already cached and the spec has not changed
+	}
+
+	// update the cache
+	c.statusCollectorNameToSpec[statusCollector.Name] = &statusCollector.Spec // readonly
+
+	combinedStatusIdentifiersToQueue := sets.New[util.ObjectIdentifier]()
+	// update resolutions that use the statuscollector
+	// this call cannot add an association that was not already present
+	for bindingName, resolutions := range c.bindingNameToResolutions {
+		for workloadObjectIdentifier, resolution := range resolutions {
+			if resolution.updateStatusCollector(statusCollector.Name, &statusCollector.Spec) { // true if changed
+				// evaluate ALL workstatuses associated with the (binding, workload object) pair
+				combinedStatusIdentifiersToQueue.Insert(c.evaluateWorkStatusesPerBindingLocked(bindingName,
+					sets.New(workloadObjectIdentifier), resolution.collectionDestinations,
+					workStatuses).UnsortedList()...)
+			}
+		}
+	}
+
+	return combinedStatusIdentifiersToQueue
+}
+
+// ResolutionExists returns true if a combinedstatus resolution exists for
+// the given binding name and workload object identifier.
+func (c *combinedStatusResolver) ResolutionExists(bindingName string, objectIdentifier util.ObjectIdentifier) bool {
+	c.RLock()
+	defer c.RUnlock()
+
+	if resolutions, exists := c.bindingNameToResolutions[bindingName]; exists {
+		_, exists = resolutions[objectIdentifier]
+		return exists
+	}
+
+	return false
+}
+
+// fetchMissingStatusCollectorSpecs fetches the missing statuscollector specs
+// from the given lister and updates the cache.
+// The method is expected to be called with the write lock held.
+func (c *combinedStatusResolver) fetchMissingStatusCollectorSpecsLocked(statusCollectorLister controllisters.StatusCollectorLister,
+	statusCollectorNames []string) {
+	for _, statusCollectorName := range statusCollectorNames {
+		if _, exists := c.statusCollectorNameToSpec[statusCollectorName]; exists {
+			continue // this method is not responsible for keeping the cache up-to-date
+		}
+
+		statusCollector, err := statusCollectorLister.Get(statusCollectorName)
+		if err != nil {
+			// fetch error should not disturb the flow.
+			// a missing spec will be reconciled when the status collector is created/updated.
+			runtime2.HandleError(fmt.Errorf("failed to get statuscollector %s: %w", statusCollectorName, err))
+		}
+
+		c.statusCollectorNameToSpec[statusCollectorName] = &statusCollector.Spec // readonly
+	}
+}
+
+// evaluateWorkStatusesPerBindingLocked evaluates workstatuses associated with the given
+// workload identifiers and destinations. Workstatuses that are not associated
+// with the two are skipped.
+// The returned set contains the identifiers of combinedstatus objects that
+// should be queued for syncing.
+// The method is expected to be called with the read lock held.
```

Comment:
Perhaps the naming convention should be to end with "ReadLocked" or "WriteLocked".